### DISCUSSION TITLE
[Deployed] Changing the main source of channels data

### DISCUSSION THREAD

**Discussion by freearhey** [iptv-org/community#103] [Announcements]
This month I plan to update the script responsible for playlist generation, after which all descriptions for channels will be loaded from the new repository [iptv-org/api](https://github.com/iptv-org/api). In repository `iptv-org/iptv` will remain only links to streams, their title, status and resolution:

```
#EXTINF:-1 tvg-id="ABSTV.ag",ABS TV (1080p) [Not 24/7]
https://query-streamlink.herokuapp.com/iptv-query?streaming-ip=https://www.twitch.tv/absliveantigua3/
```

All other information (country, language, logo etc) will be loaded automatically through new API: https://github.com/iptv-org/api#channels

```
[
  ...
  {
    "id": "KSTSDT3.us",
    "name": "KSTS-DT3",
    "network": "NBC",
    "country": "US",
    "subdivision": "US-CA",
    "city": "San Francisco",
    "broadcast_area": [
      "s/US-CA"
    ],
    "languages": [
      "eng"
    ],
    "categories": [
      "general"
    ],
    "is_nsfw": false,
    "logo": "https://cdn.tvpassport.com/image/station/100x100/nbc.png"
  },
  ...
]
```

For this to work, it is enough for the link to have a correct `tvg-id`. Full list of currently supported channels (over 24,000) and their IDs can be found here: https://iptv-org.github.io

To update the description of any channel or add a new one, you just need to edit the file [channels.csv](https://github.com/iptv-org/database/blob/master/data/channels.csv) in `iptv-org/database` repo.

<img width="608" alt="channels csv" src="https://user-images.githubusercontent.com/7253922/152466456-e91ece48-6753-47a0-8b77-b48da559172a.png">

It can be done using basic text editor or any spreadsheet editor (such as [Google Sheets](https://www.google.com/sheets/about/), [LibreOffice](https://www.libreoffice.org/discover/libreoffice/), ...).

**IMPORTANT**: this update will affect only internal playlists (files located in the `/channels` folder), public playlists should not be affected.

If you have any questions or suggestions, please leave them below.

**Comment by pvonmoradi** [comment]
@freearhey 
1- Where is the helper script that generates `streams.json` file?
2- Can a `closed_caption` or `subtitle` field be added to `streams` schema (alongside `status`, `width/height` and `bitrate`)?

**Reply by freearhey** [reply]
1- The file is generated by the [commands/database/export.js](https://github.com/iptv-org/iptv/blob/master/scripts/commands/database/export.js) script during the daily playlist update and then [copied](https://github.com/iptv-org/iptv/blob/master/.github/workflows/auto-update.yml#L125-L136) to the `iptv-org/api` repository.
2- I think so. We use [ffprobe](https://ffmpeg.org/ffprobe.html) to get info about the stream, so if the subtitle link can be obtained through it there should be no problem. I'll look into it more closely as soon as I have some free time.

**Reply by freearhey** [reply]
I did some research and found out that ffmpeg can't recognise the subtitle format used in HLS playlists: https://trac.ffmpeg.org/ticket/7337

<img width="968" alt="image" src="https://user-images.githubusercontent.com/7253922/172055191-18f3a499-311c-48c6-961b-e76e1c5cca7a.png">

But if you find a stream with subtitles that works at least in VLC player let me know and I'll try to do something about it.

**Reply by pvonmoradi** [reply]
@freearhey 
Consider this stream: https://2-fss-2.streamhoster.com/pl_138/amlst:201814-1291584/playlist.m3u8

If you do a 
```shell
ffprobe -show_streams -show_format \
                      -print_format json -loglevel quiet \
                      "https://2-fss-2.streamhoster.com/pl_138/amlst:201814-1291584/playlist.m3u8" \
> /tmp/info.json
```
Then you can find those streams that have `closed_captions` via:
```shell
jq '.streams[] | select((.closed_captions==1))' /tmp/info.json
```
or for boolean answer:
```shell
jq -e '[.streams[] | select((.closed_captions==1))] | length > 0' /tmp/info.json
```

Also another idea. Prior to switching to the [new way](https://github.com/iptv-org/api), there was a single json file `channels.json` that contained channel info + streams. It was a better endpoint for consumers IMO.
It would be helpful if we can integrate the current `streams.json` and `channels.json` into a single json file. Also, former version contained full name of languages and countries which are truncated in new api.

**Reply by freearhey** [reply]
> jq '.streams[] | select((.closed_captions==1))' /tmp/info.json

The problem is that `closed_captions` is `0` in the output data of all streams. Although in this case VLC does see the subtitles.

As for the idea to change the API, it would be better to publish it in a separate thread so other community members can see it too and comment on it.

**Reply by freearhey** [reply]
And here is a list of all the [countries](https://github.com/iptv-org/api#countries) and [languages](https://github.com/iptv-org/api#languages) with their codes and full names.

**Reply by freearhey** [reply]
Unfortunately, I didn't manage to pull subtitles from the stream with ffprobe. Besides the problem with subtitle format support I mentioned above, ffmpeg seems to show incorrect number of subtitles in the stream: https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=&submitter=&state=*&q=closed_captions&archive=both&delegate=

So at this moment adding subtitle info to API seems to be not possible. But if something changes we'll come back to this issue.
